File: impl/src/utils/person/createPerson.ts
import {JsonUtil} from "../util";
import {JsonSchema, SelectPersonQueryType} from "../JsonSchema";
import {PersonQueryType} from "./getPersonProperty";
import {JsonQueryType} from "../../JsonQueryType";
import {Position} from "./selectPerson";

export type PersonQueryElement = JsonSchema["children"]["data"]["children"]["people"]["children"]["person"];

const createNewPerson = (jsonUtil: JsonUtil, selectPerson: SelectPersonQueryType) => {
  const ruleGroup = jsonUtil.getRuleGroups();

  const raceList = ruleGroup.flatMap(ruleGroup => ruleGroup.queryAllOptional("race_rule"))
    .flatMap(e => e.queryAll("entry"));


  const selectRace = selectPerson.queryOptional("race")?.attributeMap?.race_rule_ref;

  const race = raceList.find(race => race.attributeMap.id === selectRace)
    || jsonUtil.randomFromArray(raceList)

  const name = jsonUtil.name.calculateNameFromRefString(race?.queryOptional("name")?.attributeMap?.name_rule_ref);

  let people = jsonUtil.json.queryOptional("data").queryOptional("people");
  if (!people) {
    jsonUtil.json.queryOptional("data").appendChild("people", undefined, {});
    people = jsonUtil.json.queryOptional("data").queryOptional("people");
  }
  const person = people.appendChild("person", undefined, {
    id: jsonUtil.getNextId()
  });
  console.log(`createNewPerson name ${name}`);
  if (name) {
    person.attributeMap.name = name;
  }
  if (race) {
    person.appendChild("race", undefined, {
      race_rule_ref: race.attributeMap.id
    });
  }
  person.appendChild("properties", undefined, {});
  person.appendChild("classifications", undefined, {});

  return person;
}


const applyClassification = (jsonUtil: JsonUtil, classificationRef: string, person: PersonQueryType) => {
  const ruleGroup = jsonUtil.getRuleGroups();
  const classification = ruleGroup.flatMap(ruleGroup => ruleGroup.queryAllOptional("classification_rule"))
    .flatMap(e => e.queryAllOptional("entry"))
    .find(e => e.attributeMap.id === classificationRef);

  if (!classification) {
    return person;
  }

  const classificationsElement = person.queryOrAppend("classifications");
  classificationsElement.appendChild("classification", undefined, {
    classification_rule_ref: classification.attributeMap.id,
  })


  classification.queryAllOptional("property")
    .forEach(classification => {
      const propertyRef = classification.attributeMap.property_rule_ref;
      const classificationValue = Number(jsonUtil.computeOperationFromParent(classification.query("operation")))
      const property = Number(jsonUtil.person.getProperty(person, propertyRef));
      let computedValue = property;
      switch (classification.attributeMap.is) {
        case "lessThan": {
          computedValue = Math.min(classificationValue - 1, property);
          break;
        }
        case "lessThanOrEqual": {
          computedValue = Math.min(classificationValue, property);
          break;
        }
        case "greaterThan": {
          computedValue = Math.max(classificationValue + 1, property);
          break;
        }
        case "greaterThanOrEqual": {
          computedValue = Math.max(classificationValue, property);
          break;
        }
        case "equal": {
          computedValue = classificationValue;
          break;
        }
        default: {
          throw new Error(`Unknown operation ${(classification.attributeMap.is as JsonQueryType).attributeMap.is}`);
        }
      }
      jsonUtil.person.setProperty(person, propertyRef, String(computedValue))
    })

  return person;
}

const applyProperty = (jsonUtil: JsonUtil, selectPerson: SelectPersonQueryType, person: PersonQueryType): void => {
  selectPerson.queryAllOptional("property").map(property => {
    console.log(`applyProperty ${property.getPath()}`)
    jsonUtil.person.getProperty(person, property.attributeMap.property_rule_ref);

    console.log("applyProperty has stringValue")
    let max = Number(jsonUtil.computeOperationFromParent(property.queryOptional("max")));
    let min = Number(jsonUtil.computeOperationFromParent(property.queryOptional("min")))

    if (!max) {
      max = min;
    }
    console.log(`applyProperty has max ${max}`)
    if (!min) {
      min = max;
    }
    console.log(`applyProperty has min ${min}`)
    if (!min && !max) {
      return;
    }
    console.log("applyProperty has any")
    let value = jsonUtil.randomBetweenInt(min, max);
    jsonUtil.person.setProperty(person, property.attributeMap.property_rule_ref, String(value));
  })
}


export const createPerson = (jsonUtil: JsonUtil, selectPerson: SelectPersonQueryType): PersonQueryElement => {
  try {
    console.log(`createPerson from ${selectPerson.getPath()}`)
    let person = createNewPerson(jsonUtil, selectPerson);
    selectPerson.queryAllOptional("classification").forEach(value => {
      person = applyClassification(jsonUtil, value.attributeMap.classification_rule_ref, person)
    })
    applyProperty(jsonUtil, selectPerson, person);
    return person;
  } catch (e) {
    const newError = new Error(`createPerson failed for ${selectPerson.getPath()}`);
    newError.stack += '\nCaused by: ' + e.stack;
    throw newError;
  }
}

File: implementation/src/main/java/ro/anud/xml_xsd/implementation/service/person/util/CreatePerson.java
package ro.anud.xml_xsd.implementation.service.person.util;

import ro.anud.xml_xsd.implementation.model.Type_personSelection.Classification.Classification;
import ro.anud.xml_xsd.implementation.model.Type_personSelection.Property.Property;
import ro.anud.xml_xsd.implementation.model.WorldStep.Data.People.Person.Person;
import ro.anud.xml_xsd.implementation.model.WorldStep.RuleGroup.ClassificationRule.ClassificationRule;
import ro.anud.xml_xsd.implementation.model.WorldStep.RuleGroup.PropertyRule.PropertyRule;
import ro.anud.xml_xsd.implementation.model.WorldStep.RuleGroup.RuleGroup;
import ro.anud.xml_xsd.implementation.model.WorldStep.WorldStep;
import ro.anud.xml_xsd.implementation.model.interfaces.IType_personSelection.IType_personSelection;
import ro.anud.xml_xsd.implementation.service.Mutation;
import ro.anud.xml_xsd.implementation.service.WorldStepInstance;

import static ro.anud.xml_xsd.implementation.util.LocalLogger.logEnter;
import static ro.anud.xml_xsd.implementation.util.logging.LogScope.logScope;

public class CreatePerson {
    public static Mutation<Person> createPerson(
        WorldStepInstance worldStepInstance,
        IType_personSelection<?> typePersonSelection) {
        try (var logger = logScope()){
            logger.log("creating");
            var person = createNewPerson(worldStepInstance, typePersonSelection);

            logger.log("applyProperty");
            applyProperty(worldStepInstance, typePersonSelection, person);

            logger.log("applyClassifications");
            typePersonSelection.streamClassification().forEach(classification -> {
                applyClassification(worldStepInstance, classification, person);
            });
            return Mutation.of(outInstance -> {

                return outInstance.person.repository.getOrCreate(person);

            });
        }
    }

    private static void applyProperty(
        final WorldStepInstance worldStepInstance,
        final IType_personSelection<?> typePersonSelection,
        final Person person) {
        try (var logger = logScope()){
            typePersonSelection.streamProperty()
                    .forEach(property -> {
                        try (var innerLogger = logScope("personSelection", property.getPropertyRuleRef())) {
                            innerLogger.log("compute max");
                            var maxOptional = worldStepInstance.computeOperation(property.getMax(), person);
                            innerLogger.log("compute min");
                            var minOptional = worldStepInstance.computeOperation(property.getMin(), person);

                            if (minOptional.isEmpty() && maxOptional.isEmpty()) {
                                innerLogger.log("min and max are empty");
                                worldStepInstance.person.getProperty(person, property.getPropertyRuleRef());
                                return;
                            }
                            var min = minOptional.orElse(maxOptional.orElse(0));
                            var max = maxOptional.orElse(minOptional.orElse(0));
                            innerLogger.log("min:", min, "max:", max);
                            int value = worldStepInstance.randomBetweenInt(min, max);
                            innerLogger.log("setting value:", value);
                            person.getPropertiesOrDefault()
                                    .addProperty(new ro.anud.xml_xsd.implementation.model.WorldStep.Data.People.Person.Properties.Property.Property()
                                            .setPropertyRuleRef(property.getPropertyRuleRef())
                                            .setValue(value));

                        }

                    });
            var propertyRuleRefList = typePersonSelection.streamProperty().map(Property::getPropertyRuleRef).toList();
            worldStepInstance.streamWorldStep()
                    .flatMap(WorldStep::streamRuleGroup)
                    .flatMap(RuleGroup::streamPropertyRule)
                    .flatMap(PropertyRule::streamEntry)
                    .forEach(entry -> {
                        entry.getPersonDefault().ifPresent(personDefault -> {
                            try (var innerLogger = logScope()){
                                if (propertyRuleRefList.contains(entry.getId())) {
                                    innerLogger.log("already has", entry.getId());
                                    return;
                                }
                                worldStepInstance.computeOperation(personDefault, person)
                                        .ifPresent(integer -> {
                                            innerLogger.log("setting computed value", integer);
                                            person.getPropertiesOrDefault()
                                                    .addProperty(new ro.anud.xml_xsd.implementation.model.WorldStep.Data.People.Person.Properties.Property.Property()
                                                            .setValue(integer)
                                                            .setPropertyRuleRef(entry.getId())
                                                    );
                                        });
                            }
                        });
                    });
        }

    }

    private static void applyClassification(
        final WorldStepInstance worldStepInstance,
        final Classification classification,
        final Person person) {
        try (var logger = logScope("classification:", classification.getClassificationRuleRef(), "person", person.getId())){
            var classificationRuleOptional = worldStepInstance.streamWorldStep()
                    .flatMap(WorldStep::streamRuleGroup)
                    .flatMap(RuleGroup::streamClassificationRule)
                    .flatMap(ClassificationRule::streamEntry)
                    .filter(entry -> entry.getId().equals(classification.getClassificationRuleRef()))
                    .findAny();
            if (classificationRuleOptional.isEmpty()) {
                logger.log("classificationRule not found");
                return;
            }
            var classificationRule = classificationRuleOptional.get();
            person.getClassificationsOrDefault().addClassification(new ro.anud.xml_xsd.implementation.model.WorldStep.Data.People.Person.Classifications.Classification.Classification()
                    .setClassificationRuleRef(classificationRule.getId())
            );
            classificationRule.streamProperty()
                    .forEach(property -> {
                        try (var innerLogger = logScope("property", property.getPropertyRuleRef())){
                            var propertyRef = property.getPropertyRuleRef();
                            int classificationValue = worldStepInstance.computeOperation(property.getOperation(), person).orElse(0);
                            int propertyValue = worldStepInstance.person.getProperty(person, propertyRef).orElse(0);

                            innerLogger.logTodo("remove rawNode.getAttributeRequired(\"is\")");
                            var computedValue = switch (property.rawNode().getAttributeRequired("is")) {
                                case "lessThan": {
                                    innerLogger.log("lessThan");
                                    yield Math.min(classificationValue - 1, propertyValue);
                                }
                                case "lessThanOrEqual": {
                                    innerLogger.log("lessThanOrEqual");
                                    yield Math.min(classificationValue, propertyValue);
                                }
                                case "greaterThan": {
                                    innerLogger.log("greaterThan");
                                    yield Math.max(classificationValue + 1, propertyValue);
                                }
                                case "greaterThanOrEqual": {
                                    innerLogger.log("greaterThanOrEqual");
                                    yield Math.max(classificationValue, propertyValue);
                                }
                                case "equal": {
                                    innerLogger.log("equal");
                                    yield classificationValue;
                                }
                                default:
                                    throw new RuntimeException(
                                            "Unknown operation " + property.rawNode().getAttributeRequired("is")
                                    );
                            };
                            innerLogger.log("setting propertyRef", propertyRef, "to ", computedValue);
                            worldStepInstance.person.mutateProperty(person, propertyRef, (ignored) -> computedValue);
                        }
                    });
        }

    }

    private static Person createNewPerson(
        final WorldStepInstance worldStepInstance,
        final IType_personSelection<?> typePersonSelection) {
        var logger = logEnter();
        var person = new Person();

        person.setId(worldStepInstance.getNextId());

        logger.log("adding person with id", person.getId());
        logger.logTodo("remove people.addPerson(person)");
        return person;
    }
}
